lib/ruby_lsp/store.rb

```diff
@@ -56,10 +55,13 @@ def delete(uri)
           uri: String,
           request_name: Symbol,
           block: T.proc.params(document: Document).returns(T.type_parameter(:T))
-        ).returns(T.type_parameter(:T))
+        ).returns(T.nilable(T.type_parameter(:T)))
     end
     def cache_fetch(uri, request_name, &block)
-      get(uri).cache_fetch(request_name, &block)
+      document = get(uri)
+      return if document.nil? || document.tree.nil?
+
+      document.cache_fetch(request_name, &block)
```

Comment:
I gave it a try in another branch. Here's the [comparison](https://github.com/Shopify/ruby-lsp/compare/224-alternative?expand=1)

Reply:
I'm not sure I love having to return early on every request rather than having `cache_fetch` handle this. This method now only returns `nil` if the document is not parsable.

Every other return is the result - either the cached one or the one we get by executing the block.